Test failures on Python 3.5

**Describe the bug**
Some tests are failing on Python 3.5 (but not 3.6 or 3.7).  Examples are in this build:
https://travis-ci.org/tinosulzer/PyBaMM/jobs/503052350

**To Reproduce**
Steps to reproduce the behaviour:
python3.5 tests/test_discretisations/test_discretisation.py

There's still one test failure:
```
======================================================================
FAIL: test_process_model_ode (__main__.TestDiscretise)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_discretisations/test_discretisation.py", line 384, in test_process_model_ode
    8 * np.ones_like(mesh["negative electrode"].nodes),
  File "/home/fergus/GitRepos/PyBaMM/venv35/lib/python3.5/site-packages/numpy/testing/_private/utils.py", line 896, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/fergus/GitRepos/PyBaMM/venv35/lib/python3.5/site-packages/numpy/testing/_private/utils.py", line 819, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatch: 44.4%
Max absolute difference: 3.
Max relative difference: 1.5
 x: array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,
       5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,
       5., 5., 5., 5., 5., 5., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,...
 y: array([2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,
       2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2.,...

----------------------------------------------------------------------
```

I don't understand what's going on here...

--- a/.travis.yml
+++ b/.travis.yml
@@ -29,10 +29,10 @@ matrix:
         #   env:
         #     - PYBAMM_UNIT=true
         #   if: type != cron
-        # - python: "3.5"
-        #   env:
-        #     - PYBAMM_UNIT=true
-        #   if: type != cron
+        - python: "3.5"
+          env:
+            - PYBAMM_UNIT=true
+          if: type != cron
         - python: "3.6"
           env:
             - PYBAMM_UNIT=true

--- a/.travis.yml
+++ b/.travis.yml
@@ -60,6 +60,20 @@ matrix:
                 - python3.6-dev
           env:
             - PYBAMM_UNIT=true
+        - python: "3.6"
+          addons:
+            apt:
+              sources:
+                - deadsnakes
+              packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
+                - python3.6-dev
+          env:
+            - PYBAMM_UNIT=true
             - PYBAMM_BOOKS=true
           if: type != cron
         #- python: "3.7"
@@ -141,8 +155,7 @@ install:
   - pip install pip==18
   - pip install .
   - if [[ $PYBAMM_DOCS == true ]]; then pip install -e .[docs]; fi;
-  - if [[ $PYBAMM_STYLE == true ]]; then pip install -e .[dev]; fi;
-  - if [[ $TRAVIS_EVENT_TYPE == 'cron' ]]; then pip install -e .[dev]; fi;
+  - if [[ $PYBAMM_STYLE == true || $PYBAMM_BOOKS ]]; then pip install -e .[dev]; fi;
   - if [[ $PYBAMM_COVER == true ]]; then pip install coverage codecov; fi;
   - source scripts/install_scikits_odes.sh
 

--- a/.travis.yml
+++ b/.travis.yml
@@ -66,7 +66,7 @@ matrix:
             - PYBAMM_STYLE=true
           if: type != cron
         - python: "3.6"
-           addons:
+          addons:
             apt:
               packages:
               - python3.6-dev

--- a/.travis.yml
+++ b/.travis.yml
@@ -34,7 +34,7 @@ matrix:
           addons:
             apt:
               packages:
-              - python3.5-dev
+                - python3.5-dev
           env:
             - PYBAMM_UNIT=true
           if: type != cron
@@ -42,7 +42,7 @@ matrix:
           addons:
             apt:
               packages:
-              - python3.6-dev
+                - python3.6-dev
           env:
             - PYBAMM_UNIT=true
           if: type != cron
@@ -69,7 +69,7 @@ matrix:
           addons:
             apt:
               packages:
-              - python3.6-dev
+                - python3.6-dev
           env:
             - PYBAMM_COVER=true
           if: type != cron
@@ -78,7 +78,7 @@ matrix:
           addons:
             apt:
               packages:
-              - python3.6-dev
+                - python3.6-dev
           if: type == cron
 
 # Install dependencies

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,7 +44,6 @@ matrix:
                 - python3.5-dev
           env:
             - PYBAMM_UNIT=true
-            - PYBAMM_BOOKS=true
           if: type != cron
         - python: "3.6"
           addons:
@@ -74,7 +73,6 @@ matrix:
                 - python3.6-dev
           env:
             - PYBAMM_UNIT=true
-            - PYBAMM_BOOKS=true
           if: type != cron
         #- python: "3.7"
         #  env:
@@ -145,6 +143,8 @@ matrix:
                 - liblapack-dev
                 - graphviz
                 - python3.6-dev
+          env:
+            - PYBAMM_BOOKS=true
           if: type == cron
 
 # Install dependencies

--- a/.travis.yml
+++ b/.travis.yml
@@ -33,18 +33,34 @@ matrix:
         - python: "3.5"
           addons:
             apt:
+              sources:
+                - deadsnakes
               packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
                 - python3.5-dev
           env:
             - PYBAMM_UNIT=true
+            - PYBAMM_BOOKS=true
           if: type != cron
         - python: "3.6"
           addons:
             apt:
+              sources:
+                - deadsnakes
               packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
                 - python3.6-dev
           env:
             - PYBAMM_UNIT=true
+            - PYBAMM_BOOKS=true
           if: type != cron
         #- python: "3.7"
         #  env:
@@ -58,17 +74,46 @@ matrix:
         #   if: type != cron
         # Docs, style and cover checking, latest Python version only
         - python: "3.6"
+          addons:
+            apt:
+              sources:
+                - deadsnakes
+              packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
+                - python3.6-dev
           env:
             - PYBAMM_DOCS=true
           if: type != cron
         - python: "3.6"
+          addons:
+            apt:
+              sources:
+                - deadsnakes
+              packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
+                - python3.6-dev
           env:
             - PYBAMM_STYLE=true
           if: type != cron
         - python: "3.6"
           addons:
             apt:
+              sources:
+                - deadsnakes
               packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
                 - python3.6-dev
           env:
             - PYBAMM_COVER=true
@@ -77,7 +122,14 @@ matrix:
         - python: "3.6"
           addons:
             apt:
+              sources:
+                - deadsnakes
               packages:
+                - gfortran
+                - gcc
+                - libopenblas-dev
+                - liblapack-dev
+                - graphviz
                 - python3.6-dev
           if: type == cron
 
@@ -106,7 +158,7 @@ script:
   - if [[ $PYBAMM_DOCS == true ]]; then python run-tests.py --doctest; fi;
   - if [[ $PYBAMM_STYLE == true ]]; then python -m flake8; fi;
   - if [[ $PYBAMM_COVER == true ]]; then coverage run run-tests.py --nosub; fi;
-  - if [[ $TRAVIS_EVENT_TYPE == 'cron' ]]; then travis_wait 120 python run-tests.py --books; fi;
+  - if [[ $PYBAMM_BOOKS == true ]]; then travis_wait 120 python run-tests.py --books; fi;
 
 after_success:
   - if [[ $PYBAMM_COVER == true ]]; then codecov; fi;

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,8 +6,9 @@ language: python
 
 addons:
   apt:
+    sources:
+    - deadsnakes
     packages:
-    - python3-dev
     - gfortran
     - gcc
     - libopenblas-dev
@@ -30,10 +31,18 @@ matrix:
         #     - PYBAMM_UNIT=true
         #   if: type != cron
         - python: "3.5"
+          addons:
+            apt:
+              packages:
+              - python3.5-dev
           env:
             - PYBAMM_UNIT=true
           if: type != cron
         - python: "3.6"
+          addons:
+            apt:
+              packages:
+              - python3.6-dev
           env:
             - PYBAMM_UNIT=true
           if: type != cron
@@ -57,11 +66,19 @@ matrix:
             - PYBAMM_STYLE=true
           if: type != cron
         - python: "3.6"
+           addons:
+            apt:
+              packages:
+              - python3.6-dev
           env:
             - PYBAMM_COVER=true
           if: type != cron
           # Cron jobs
         - python: "3.6"
+          addons:
+            apt:
+              packages:
+              - python3.6-dev
           if: type == cron
 
 # Install dependencies

--- a/pybamm/discretisations/discretisation.py
+++ b/pybamm/discretisations/discretisation.py
@@ -169,8 +169,8 @@ class Discretisation(object):
 
         # Concatenate initial conditions into a single vector
         # check that all initial conditions are set
-        processed_concatenated_initial_conditions = self._concatenate_init(
-            processed_initial_conditions
+        processed_concatenated_initial_conditions = self._concatenate_in_order(
+            processed_initial_conditions, check_complete=True
         ).evaluate(0, None)
 
         return processed_initial_conditions, processed_concatenated_initial_conditions
@@ -195,12 +195,17 @@ class Discretisation(object):
         # Discretise right-hand sides, passing domain from variable
         processed_rhs = self.process_dict(model.rhs)
         # Concatenate rhs into a single state vector
-        processed_concatenated_rhs = self.concatenate(*processed_rhs.values())
+        # Need to concatenate in order as the ordering of equations could be different
+        # in processed_rhs and model.rhs (for Python Version <= 3.5)
+        processed_concatenated_rhs = self._concatenate_in_order(processed_rhs)
 
         # Discretise and concatenate algebraic equations
         processed_algebraic = self.process_dict(model.algebraic)
-        processed_concatenated_algebraic = self.concatenate(
-            *processed_algebraic.values()
+
+        # Need to concatenate in order as the ordering of equations could be different
+        # in processed_rhs and model.rhs (for Python Version <= 3.5)
+        processed_concatenated_algebraic = self._concatenate_in_order(
+            processed_algebraic
         )
 
         return (
@@ -404,10 +409,9 @@ class Discretisation(object):
     def concatenate(self, *symbols):
         return pybamm.NumpyConcatenation(*symbols)
 
-    def _concatenate_init(self, var_eqn_dict):
+    def _concatenate_in_order(self, var_eqn_dict, check_complete=False):
         """
-        Concatenate a dictionary of {variable: equation} initial conditions using
-        self._y_slices
+        Concatenate a dictionary of {variable: equation} using self._y_slices
 
         The keys/variables in `var_eqn_dict` must be the same as the ids in
         `self._y_slices`.
@@ -432,14 +436,16 @@ class Discretisation(object):
                 unpacked_variables.extend([var for var in symbol.children])
             else:
                 unpacked_variables.append(symbol)
-        # Check keys from the given var_eqn_dict against self._y_slices
-        ids = {v.id for v in unpacked_variables}
-        if ids != self._y_slices.keys():
-            given_variable_names = [v.name for v in var_eqn_dict.keys()]
-            raise pybamm.ModelError(
-                "Initial conditions are insufficient. Only "
-                "provided for {} ".format(given_variable_names)
-            )
+
+        if check_complete:
+            # Check keys from the given var_eqn_dict against self._y_slices
+            ids = {v.id for v in unpacked_variables}
+            if ids != self._y_slices.keys():
+                given_variable_names = [v.name for v in var_eqn_dict.keys()]
+                raise pybamm.ModelError(
+                    "Initial conditions are insufficient. Only "
+                    "provided for {} ".format(given_variable_names)
+                )
 
         equations = list(var_eqn_dict.values())
         slices = [self._y_slices[var.id] for var in unpacked_variables]

--- a/pybamm/discretisations/discretisation.py
+++ b/pybamm/discretisations/discretisation.py
@@ -195,6 +195,7 @@ class Discretisation(object):
         # Discretise right-hand sides, passing domain from variable
         processed_rhs = self.process_dict(model.rhs)
         # Concatenate rhs into a single state vector
+
         # Need to concatenate in order as the ordering of equations could be different
         # in processed_rhs and model.rhs (for Python Version <= 3.5)
         processed_concatenated_rhs = self._concatenate_in_order(processed_rhs)
@@ -202,8 +203,6 @@ class Discretisation(object):
         # Discretise and concatenate algebraic equations
         processed_algebraic = self.process_dict(model.algebraic)
 
-        # Need to concatenate in order as the ordering of equations could be different
-        # in processed_rhs and model.rhs (for Python Version <= 3.5)
         processed_concatenated_algebraic = self._concatenate_in_order(
             processed_algebraic
         )
@@ -235,8 +234,23 @@ class Discretisation(object):
         # diagonal mass matrix for the model
         mass_list = []
 
+        # get a list of model rhs variables that are sorted according to
+        # where they are in the state vector
+        model_variables = model.rhs.keys()
+        model_slices = []
+        for v in model_variables:
+            if isinstance(v, pybamm.Concatenation):
+                model_slices.append(
+                    slice(self._y_slices[v.children[0].id].start,
+                          self._y_slices[v.children[-1].id].stop)
+                )
+            else:
+                model_slices.append(self._y_slices[v.id])
+        sorted_model_variables = [v for _, v in
+                                  sorted(zip(model_slices, model_variables))]
+
         # Process mass matrices for the differential equations
-        for var in model.rhs.keys():
+        for var in sorted_model_variables:
             if var.domain == []:
                 # If variable domain empty then mass matrix is just 1
                 mass_list.append(1.0)
@@ -431,11 +445,17 @@ class Discretisation(object):
         """
         # Unpack symbols in variables that are concatenations of variables
         unpacked_variables = []
+        slices = []
         for symbol in var_eqn_dict.keys():
             if isinstance(symbol, pybamm.Concatenation):
                 unpacked_variables.extend([var for var in symbol.children])
+                slices.append(
+                    slice(self._y_slices[symbol.children[0].id].start,
+                          self._y_slices[symbol.children[-1].id].stop)
+                )
             else:
                 unpacked_variables.append(symbol)
+                slices.append(self._y_slices[symbol.id])
 
         if check_complete:
             # Check keys from the given var_eqn_dict against self._y_slices
@@ -448,7 +468,6 @@ class Discretisation(object):
                 )
 
         equations = list(var_eqn_dict.values())
-        slices = [self._y_slices[var.id] for var in unpacked_variables]
 
         # sort equations according to slices
         sorted_equations = [eq for _, eq in sorted(zip(slices, equations))]

--- a/pybamm/meshes/submeshes.py
+++ b/pybamm/meshes/submeshes.py
@@ -39,8 +39,13 @@ class Uniform1DSubMesh(SubMesh1D):
         # currently accept lims and npts as dicts. This may get changed at a future
         # date depending on the form of mesh we desire for 2D/3D
 
-        spatial_lims = list(lims.values())[0]
-        npts = list(npts.values())[0]
+        # check that only one variable passed in
+        if len(lims) != 1:
+            raise ValueError("lims should only contain a single variable")
+
+        var = list(lims.keys())[0]
+        spatial_lims = lims[var]
+        npts = npts[var.id]
 
         edges = np.linspace(spatial_lims["min"], spatial_lims["max"], npts + 1)
 

--- a/pybamm/parameters/parameter_values.py
+++ b/pybamm/parameters/parameter_values.py
@@ -145,9 +145,9 @@ class ParameterValues(dict):
         self.process_model(model, processing="update")
 
         # update discretised quantities using disc
-        model.concatenated_rhs = disc.concatenate(*model.rhs.values())
-        model.concatenated_algebraic = disc.concatenate(*model.algebraic.values())
-        model.concatenated_initial_conditions = disc._concatenate_init(
+        model.concatenated_rhs = disc._concatenate_in_order(model.rhs)
+        model.concatenated_algebraic = disc._concatenate_in_order(model.algebraic)
+        model.concatenated_initial_conditions = disc._concatenate_in_order(
             model.initial_conditions
         ).evaluate(0, None)
 

--- a/pybamm/solvers/scikits_dae_solver.py
+++ b/pybamm/solvers/scikits_dae_solver.py
@@ -7,7 +7,7 @@ import pybamm
 
 import numpy as np
 import importlib
-from scipy.sparse import issparse
+import scipy.sparse as sparse
 
 scikits_odes_spec = importlib.util.find_spec("scikits")
 if scikits_odes_spec is not None:
@@ -15,8 +15,6 @@ if scikits_odes_spec is not None:
     if scikits_odes_spec is not None:
         scikits_odes = importlib.util.module_from_spec(scikits_odes_spec)
         scikits_odes_spec.loader.exec_module(scikits_odes)
-else:
-    jac_class = object
 
 
 class ScikitsDaeSolver(pybamm.DaeSolver):
@@ -81,10 +79,20 @@ class ScikitsDaeSolver(pybamm.DaeSolver):
         extra_options = {"old_api": False, "rtol": self.tol, "atol": self.tol}
 
         if jacobian:
-            # Put the user-supplied Jacobian into the SUNDIALS Class
-            jacfn = JacobianFunctionIDA()
-            jacfn.set_jacobian(mass_matrix, jacobian)
-            extra_options.update({"jacfn": jacfn})
+            jac_y0_t0 = jacobian(t_eval[0], y0)
+            if sparse.issparse(jac_y0_t0):
+                def jacfn(t, y, ydot, residuals, cj, J):
+                    jac_eval = jacobian(t, y) - cj * mass_matrix
+                    J[:][:] = jac_eval.toarray()
+
+            else:
+                def jacfn(t, y, ydot, residuals, cj, J):
+                    jac_eval = jacobian(t, y) - cj * mass_matrix
+                    J[:][:] = jac_eval
+
+            extra_options.update({
+                "jacfn": jacfn
+            })
 
         if events:
             extra_options.update({"rootfn": rootfn, "nr_rootfns": len(events)})
@@ -98,30 +106,3 @@ class ScikitsDaeSolver(pybamm.DaeSolver):
 
         # return solution, we need to tranpose y to match scipy's interface
         return sol.values.t, np.transpose(sol.values.y)
-
-
-class JacobianFunctionIDA(jac_class):
-    def set_jacobian(self, mass_matrix, jacobian):
-        """
-        Sets the user supplied mass matrix and Jacobian function for the DAE model.
-
-        Parameters
-        ----------
-        mass_matrix : array_like
-            The (sparse) mass matrix for the chosen spatial method.
-        jacobian : method
-            A function that takes in t and y and returns the Jacobian.
-
-        """
-        self.mass_matrix = mass_matrix
-        self.jacobian = jacobian
-
-    def evaluate(self, t, y, ydot, residuals, cj, return_jacobian):
-        # scikits_odes requires the full (dense) jacobian
-        jac_eval = self.jacobian(t, y) - cj * self.mass_matrix
-        if issparse(jac_eval):
-            return_jacobian[:][:] = jac_eval.toarray()
-        else:
-            return_jacobian[:][:] = jac_eval
-
-        return 0

--- a/pybamm/solvers/scikits_dae_solver.py
+++ b/pybamm/solvers/scikits_dae_solver.py
@@ -15,9 +15,6 @@ if scikits_odes_spec is not None:
     if scikits_odes_spec is not None:
         scikits_odes = importlib.util.module_from_spec(scikits_odes_spec)
         scikits_odes_spec.loader.exec_module(scikits_odes)
-        from scikits.odes.sundials import ida
-
-        jac_class = ida.IDA_JacRhsFunction
 else:
     jac_class = object
 

--- a/setup.py
+++ b/setup.py
@@ -16,7 +16,7 @@ setup(
     url="https://github.com/pybamm-team/PyBaMM",
     # List of dependencies
     install_requires=[
-        "numpy>=1.14",
+        "numpy>=1.16",
         "scipy>=1.0",
         "pandas>=0.23",
         "anytree>=2.4.3",

--- a/tests/unit/test_discretisations/test_discretisation.py
+++ b/tests/unit/test_discretisations/test_discretisation.py
@@ -396,22 +396,22 @@ class TestDiscretise(unittest.TestCase):
 
         disc.process_model(model)
         y0 = model.concatenated_initial_conditions
-        np.testing.assert_array_equal(
-            y0,
-            np.concatenate(
-                [
-                    2 * np.ones_like(combined_submesh[0].nodes),
-                    5 * np.ones_like(mesh["negative electrode"][0].nodes),
-                    8 * np.ones_like(mesh["negative electrode"][0].nodes),
-                ]
-            ),
-        )
+        y0_expect = np.array([])
+        for var, init in model.initial_conditions.items():
+            if var.id == c.id:
+                vect = init.evaluate() * np.ones_like(combined_submesh[0].nodes)
+            else:
+                vect = init.evaluate() * np.ones_like(mesh[var.domain[0]][0].nodes)
+
+            y0_expect = np.concatenate([y0_expect,vect])
+
+        np.testing.assert_array_equal(y0,y0_expect)
+
         # grad and div are identity operators here
         np.testing.assert_array_equal(y0, model.concatenated_rhs.evaluate(None, y0))
-        c0, T0, S0 = np.split(
-            y0,
-            np.cumsum([combined_submesh[0].npts, mesh["negative electrode"][0].npts]),
-        )
+
+        S0 = model.initial_conditions[S].evaluate() * np.ones_like(mesh[S.domain[0]][0].nodes)
+        T0 = model.initial_conditions[T].evaluate() * np.ones_like(mesh[T.domain[0]][0].nodes)
         np.testing.assert_array_equal(S0 * T0, model.variables["ST"].evaluate(None, y0))
 
         # mass matrix is identity

--- a/tests/unit/test_discretisations/test_discretisation.py
+++ b/tests/unit/test_discretisations/test_discretisation.py
@@ -10,7 +10,7 @@ from scipy.sparse import block_diag
 
 
 class TestDiscretise(unittest.TestCase):
-    def test_concatenate_init(self):
+    def test_concatenate_in_order(self):
         a = pybamm.Variable("a")
         b = pybamm.Variable("b")
         c = pybamm.Variable("c")
@@ -25,16 +25,16 @@ class TestDiscretise(unittest.TestCase):
         disc = get_discretisation_for_testing()
 
         disc._y_slices = {c.id: slice(0, 1), a.id: slice(2, 3), b.id: slice(3, 4)}
-        result = disc._concatenate_init(initial_conditions)
+        result = disc._concatenate_in_order(initial_conditions)
 
-        self.assertIsInstance(result, pybamm.NumpyConcatenation)
+        self.assertIsInstance(result, pybamm.NumpyConcatenation, check_complete=True)
         self.assertEqual(result.children[0].evaluate(), 1)
         self.assertEqual(result.children[1].evaluate(), 2)
         self.assertEqual(result.children[2].evaluate(), 3)
 
         initial_conditions = {a: pybamm.Scalar(2), b: pybamm.Scalar(3)}
         with self.assertRaises(pybamm.ModelError):
-            result = disc._concatenate_init(initial_conditions)
+            result = disc._concatenate_in_order(initial_conditions, check_complete=True)
 
     def test_discretise_slicing(self):
         # create discretisation
@@ -403,15 +403,17 @@ class TestDiscretise(unittest.TestCase):
             else:
                 vect = init.evaluate() * np.ones_like(mesh[var.domain[0]][0].nodes)
 
-            y0_expect = np.concatenate([y0_expect,vect])
+            y0_expect = np.concatenate([y0_expect, vect])
 
-        np.testing.assert_array_equal(y0,y0_expect)
+        np.testing.assert_array_equal(y0, y0_expect)
 
         # grad and div are identity operators here
         np.testing.assert_array_equal(y0, model.concatenated_rhs.evaluate(None, y0))
 
-        S0 = model.initial_conditions[S].evaluate() * np.ones_like(mesh[S.domain[0]][0].nodes)
-        T0 = model.initial_conditions[T].evaluate() * np.ones_like(mesh[T.domain[0]][0].nodes)
+        S0 = model.initial_conditions[S].evaluate(
+        ) * np.ones_like(mesh[S.domain[0]][0].nodes)
+        T0 = model.initial_conditions[T].evaluate(
+        ) * np.ones_like(mesh[T.domain[0]][0].nodes)
         np.testing.assert_array_equal(S0 * T0, model.variables["ST"].evaluate(None, y0))
 
         # mass matrix is identity

--- a/tests/unit/test_discretisations/test_discretisation.py
+++ b/tests/unit/test_discretisations/test_discretisation.py
@@ -27,7 +27,7 @@ class TestDiscretise(unittest.TestCase):
         disc._y_slices = {c.id: slice(0, 1), a.id: slice(2, 3), b.id: slice(3, 4)}
         result = disc._concatenate_in_order(initial_conditions)
 
-        self.assertIsInstance(result, pybamm.NumpyConcatenation, check_complete=True)
+        self.assertIsInstance(result, pybamm.NumpyConcatenation)
         self.assertEqual(result.children[0].evaluate(), 1)
         self.assertEqual(result.children[1].evaluate(), 2)
         self.assertEqual(result.children[2].evaluate(), 3)
@@ -397,11 +397,13 @@ class TestDiscretise(unittest.TestCase):
         disc.process_model(model)
         y0 = model.concatenated_initial_conditions
         y0_expect = np.array([])
-        for var, init in model.initial_conditions.items():
-            if var.id == c.id:
-                vect = init.evaluate() * np.ones_like(combined_submesh[0].nodes)
+        for var_id, _ in sorted(disc._y_slices.items(), key=lambda kv: kv[1]):
+            if var_id == c.id:
+                vect = 2 * np.ones_like(combined_submesh[0].nodes)
+            elif var_id == T.id:
+                vect = 5 * np.ones_like(mesh["negative electrode"][0].nodes)
             else:
-                vect = init.evaluate() * np.ones_like(mesh[var.domain[0]][0].nodes)
+                vect = 8 * np.ones_like(mesh["negative electrode"][0].nodes)
 
             y0_expect = np.concatenate([y0_expect, vect])
 

--- a/tests/unit/test_geometry/test_geometry.py
+++ b/tests/unit/test_geometry/test_geometry.py
@@ -87,14 +87,14 @@ class TestGeometry(unittest.TestCase):
         geometry1Dmicro = pybamm.Geometry1DMicro()
         geometry = pybamm.Geometry(geometry1Dmacro, geometry1Dmicro)
         self.assertEqual(
-            list(geometry.keys()),
-            [
+            set(geometry.keys()),
+            set([
                 "negative electrode",
                 "separator",
                 "positive electrode",
                 "negative particle",
                 "positive particle",
-            ],
+            ]),
         )
 
         # update with custom geometry
@@ -117,14 +117,14 @@ class TestGeometry(unittest.TestCase):
         geometry1Dmicro = pybamm.Geometry1DMicro()
         geometry = pybamm.Geometry(geometry3Dmacro, geometry1Dmicro)
         self.assertEqual(
-            list(geometry.keys()),
-            [
+            set(geometry.keys()),
+            set([
                 "negative electrode",
                 "separator",
                 "positive electrode",
                 "negative particle",
                 "positive particle",
-            ],
+            ]),
         )
 
         with self.assertRaises(ValueError):
@@ -134,25 +134,25 @@ class TestGeometry(unittest.TestCase):
     def test_combine_geometries_strings(self):
         geometry = pybamm.Geometry("1D macro", "1D micro")
         self.assertEqual(
-            list(geometry.keys()),
-            [
+            set(geometry.keys()),
+            set([
                 "negative electrode",
                 "separator",
                 "positive electrode",
                 "negative particle",
                 "positive particle",
-            ],
+            ]),
         )
         geometry = pybamm.Geometry("3D macro", "1D micro")
         self.assertEqual(
-            list(geometry.keys()),
-            [
+            set(geometry.keys()),
+            set([
                 "negative electrode",
                 "separator",
                 "positive electrode",
                 "negative particle",
                 "positive particle",
-            ],
+            ]),
         )
 
 

--- a/tests/unit/test_solvers/test_scikits_solvers.py
+++ b/tests/unit/test_solvers/test_scikits_solvers.py
@@ -377,6 +377,8 @@ class TestScikitsSolver(unittest.TestCase):
         var2 = pybamm.Variable("var2", domain=whole_cell)
         model.rhs = {var1: var1, var2: 1 - var1}
         model.initial_conditions = {var1: 1.0, var2: -1.0}
+        model.variables = {"var1": var1, "var2": var2}
+
         disc = get_discretisation_for_testing()
         disc.process_model(model)
 
@@ -387,10 +389,18 @@ class TestScikitsSolver(unittest.TestCase):
         )
         N = combined_submesh[0].npts
 
+        # construct jacobian in order of model.rhs
+        J = []
+        for var in model.rhs.keys():
+            if var.id == var1.id:
+                J.append([np.eye(N), np.zeros((N, N))])
+            else:
+                J.append([-1.0 * np.eye(N), np.zeros((N, N))])
+
+        J = np.block(J)
+
         def jacobian(t, y):
-            return np.block(
-                [[np.eye(N), np.zeros((N, N))], [-1.0 * np.eye(N), np.zeros((N, N))]]
-            )
+            return J
 
         model.jacobian = jacobian
 
@@ -399,8 +409,16 @@ class TestScikitsSolver(unittest.TestCase):
         t_eval = np.linspace(0, 1, 100)
         solver.solve(model, t_eval)
         np.testing.assert_array_equal(solver.t, t_eval)
-        np.testing.assert_allclose(solver.y[0], np.exp(solver.t))
-        np.testing.assert_allclose(solver.y[-1], solver.t - np.exp(solver.t))
+
+        T, Y = solver.t, solver.y
+        np.testing.assert_array_almost_equal(
+            model.variables["var1"].evaluate(T, Y),
+            np.ones((N, T.size)) * np.exp(T[np.newaxis, :])
+        )
+        np.testing.assert_array_almost_equal(
+            model.variables["var2"].evaluate(T, Y),
+            np.ones((N, T.size)) * (T[np.newaxis, :] - np.exp(T[np.newaxis, :]))
+        )
 
     def test_model_solver_dae(self):
         # Create model

--- a/tests/unit/test_solvers/test_scipy_solver.py
+++ b/tests/unit/test_solvers/test_scipy_solver.py
@@ -162,6 +162,7 @@ class TestScipySolver(unittest.TestCase):
         var2 = pybamm.Variable("var2", domain=whole_cell)
         model.rhs = {var1: var1, var2: 1 - var1}
         model.initial_conditions = {var1: 1.0, var2: -1.0}
+        model.variables = {"var1": var1, "var2": var2}
 
         # create discretisation
         mesh = get_mesh_for_testing()
@@ -175,10 +176,18 @@ class TestScipySolver(unittest.TestCase):
         )
         N = combined_submesh[0].npts
 
+        # construct jacobian in order of model.rhs
+        J = []
+        for var in model.rhs.keys():
+            if var.id == var1.id:
+                J.append([np.eye(N), np.zeros((N, N))])
+            else:
+                J.append([-1.0 * np.eye(N), np.zeros((N, N))])
+
+        J = np.block(J)
+
         def jacobian(t, y):
-            return np.block(
-                [[np.eye(N), np.zeros((N, N))], [-1.0 * np.eye(N), np.zeros((N, N))]]
-            )
+            return J
 
         model.jacobian = jacobian
 
@@ -187,8 +196,16 @@ class TestScipySolver(unittest.TestCase):
         t_eval = np.linspace(0, 1, 100)
         solver.solve(model, t_eval)
         np.testing.assert_array_equal(solver.t, t_eval)
-        np.testing.assert_allclose(solver.y[0], np.exp(solver.t))
-        np.testing.assert_allclose(solver.y[-1], solver.t - np.exp(solver.t))
+
+        T, Y = solver.t, solver.y
+        np.testing.assert_array_almost_equal(
+            model.variables["var1"].evaluate(T, Y),
+            np.ones((N, T.size)) * np.exp(T[np.newaxis, :])
+        )
+        np.testing.assert_array_almost_equal(
+            model.variables["var2"].evaluate(T, Y),
+            np.ones((N, T.size)) * (T[np.newaxis, :] - np.exp(T[np.newaxis, :]))
+        )
 
 
 if __name__ == "__main__":

